apis/v1alpha2/validation/grpcroute.go

```diff
@@ -53,9 +72,166 @@ func validateGRPCRouteRules(rules []gatewayv1a2.GRPCRouteRule, path *field.Path)
 func validateRuleMatches(matches []gatewayv1a2.GRPCRouteMatch, path *field.Path) field.ErrorList {
 	var errs field.ErrorList
 	for i, m := range matches {
-		if m.Method != nil && m.Method.Service == nil && m.Method.Method == nil {
-			errs = append(errs, field.Required(path.Index(i).Child("methods"), "one or both of `service` or `method` must be specified"))
-			return errs
+		if m.Method != nil {
+			errs = append(errs, validateGRPCMethod(*m.Method, path.Index(i).Child("method"))...)
+		}
+		if m.Headers != nil {
+			errs = append(errs, validateGRPCHeaderMatches(m.Headers, path.Index(i).Child("headers"))...)
+		}
+	}
+	return errs
+}
+
+// validateGRPCMethod validates that no header name
+// is matched more than once (case-insensitive).
+func validateGRPCMethod(m gatewayv1a2.GRPCMethodMatch, path *field.Path) field.ErrorList {
+	var errs field.ErrorList
+	if m.Service == nil && m.Method == nil {
+		return append(errs, field.Required(path, "one or both of `service` or `method` must be specified"))
+	}
+
+	if m.Service != nil {
+		for _, invalidSeq := range invalidPathSequences {
+			if strings.Contains(*m.Service, invalidSeq) {
+				errs = append(errs, field.Invalid(path.Child("service"), *m.Service, fmt.Sprintf("must not contain %q", invalidSeq)))
+			}
+		}
+
+		for _, invalidSuff := range invalidPathSuffixes {
+			if strings.HasSuffix(*m.Service, invalidSuff) {
+				errs = append(errs, field.Invalid(path.Child("service"), *m.Service, fmt.Sprintf("cannot end with '%s'", invalidSuff)))
```

Comment:
should this be a constraint on the CRD instead of the webhook? I think its preferred to have as much in the CRDs as possible?

Reply:
++

Reply:
The existing regex for method is `^[^\/]*$` (blocking only slash characters), so I guess we could add some stuff in there as well. But the idea here is to be belt-and-suspenders for this.

Reply:
I've been chatting with @gnossen about the best approach here and for now have tightened up CRD based validation since it appears that a safe regex is simple enough to write. Will hold off until I can get a clear reference to GRPC spec to confirm validations I've proposed here are accurate.